solr/core/src/java/org/apache/solr/util/tracing/GlobalTracer.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util.tracing;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Random;
+
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.noop.NoopTracerFactory;
+import io.opentracing.propagation.Format;
+
+public class GlobalTracer {
+  private static final Random RANDOM;
+  static {
+    // We try to make things reproducible in the context of our tests by initializing the random instance
+    // based on the current seed
+    String seed = System.getProperty("tests.seed");
+    if (seed == null) {
+      RANDOM = new Random();
+    } else {
+      RANDOM = new Random(seed.hashCode());
+    }
+  }
+
+  private static final Tracer NOOP_TRACER = NoopTracerFactory.create();
+  private static volatile Tracer TRACER = NOOP_TRACER;
+  private static volatile int rate;
+  private final static ThreadLocal<Tracer> threadLocal = new ThreadLocal<>();
+
+  public static void setup(Tracer tracer, double rate) {
+    GlobalTracer.rate = (int) (rate * 10);
+    GlobalTracer.TRACER = tracer;
+  }
+
+  public static boolean tracing() {
+    return threadLocal.get() != null
+        && threadLocal.get() != NOOP_TRACER
+        && threadLocal.get().activeSpan() != null;
+  }
+
+  public static SpanContext extract(HttpServletRequest request) {
+    SpanContext spanContext = TRACER.extract(Format.Builtin.HTTP_HEADERS, new HttpServletCarrier(request));
+    if (spanContext != null) {
+      threadLocal.set(TRACER);
+    }
+    return spanContext;
+  }
+
+  public static Tracer get() {
+    Tracer tracer = threadLocal.get();
+    if (tracer != null)
+      return tracer;
+
+    if (traced()) {
```

Comment:
Would there be a way of propagating the trace/span ids (i.e. still doing the tracing) just not reporting it?
This way we'd have like 1/1000 samples for bigger trends in the report UI while being able to log trace ids of all requests from the MDC. (If you have an error you could still grep through all the logs even if the request wasn't sampled)

Reply:
I can't find an easy way for doing this since in OpenTracing sampleRate is handled by Tracer implementation. The only way to do that is wrapping all the apis, it will take quite effort and not necessary now, it can be solved in another issue.

Reply:
I think the trace id if it exists should be added to MDC anyway.

Reply:
If it's being handled by the Tracer implementation wouldn't it be wise not to reimplement the logic here and use the settings of e.g. Jaeger to deal with which traces are to be sampled?

Reply:
The tracing here should be able to support different backends, not just Jaeger. Therefore unify `samplePercentage` configuration in a central place is wiser than each of `TracerConfigurator` needs to implements their own way for configuring it. 

Plus `samplePercentage` is moved to `cluster properties` for supporting dynamically changing the sample rate on the fly. Therefore using Tracer implementation for setting sampling rate is not appropriate.

Reply:
When we're getting a request that is already sampled by the calling process we should continue sampling it regardless of the sampling rate set here (failing to do so would result in incomplete reports). I think that a sampling rate on its own is only half of the solution.

If we hit a node and it distributes the tracing context but other nodes ignore it 99% of the time we'd get only incomplete tracing data. That's why I would leave it to the implementor. Dealing with the thread locals could also be spared.

Reply:
> When we're getting a request that is already sampled by the calling process we should continue sampling it regardless of the sampling rate set here (failing to do so would result in incomplete reports). I think that a sampling rate on its own is only half of the solution.

We already do that in the PR. Whenever the context present in the header, we ignore sampling rate.

Reply:
One last question regarding this:
If the first node decided not to sample a trace is there a way to notify the subsequent nodes not to roll the dice again?